cmd/paths.go

```diff
@@ -256,77 +255,124 @@ func pathsFetchCmd(a *appState) *cobra.Command {
-						if err = json.Unmarshal(byt, p); err != nil {
-							cleanupDir(localRepo)
-							return fmt.Errorf("failed to unmarshal file %s: %w", pth, err)
-						}
-
-						if p.Dst.ChainID == dstChain.ChainID() {
-							pthName := strings.Split(f.Name(), ".")[0]
-							if err = a.Config.AddPath(pthName, p); err != nil {
-								return fmt.Errorf("failed to add path %s: %w", pth, err)
-							}
-
-							fmt.Fprintf(cmd.ErrOrStderr(), "added path %s...\n", pthName)
-						}
-					}
+				_, exist := a.Config.Paths[pthName]
+				if exist && !overwrite {
+					// fmt.Printf("skipping:  %s already exists in config, use -o to overwrite (clears filters) \n", pthName)
+					fmt.Fprintf(cmd.ErrOrStderr(), "skipping:  %s already exists in config, use -o to overwrite (clears filters)\n", pthName)
+					continue
+				}
 
-					if err := a.OverwriteConfig(a.Config); err != nil {
-						return err
+				// TODO: Don't use github api. Potentially use: https://github.com/eco-stake/cosmos-directory once they integrate IBC data into restAPI. This will avoid rate limits.
+				fileName := pthName + ".json"
+				regPath := path.Join("_IBC", fileName)
+				reader, _, err := client.Repositories.DownloadContents(cmd.Context(), "cosmos", "chain-registry", regPath, nil)
+				if err != nil {
+					if errors.As(err, new(*github.RateLimitError)) {
+						return fmt.Errorf("github API rate limit reached while querying data ERR: %w", err)
 					}
+					fmt.Fprintf(cmd.ErrOrStderr(), "not found:  path %s not found in repo 'cosmos/chain-registry', folder '_IBC'\n", pthName)
+					continue
+				}
+
+				b, err := io.ReadAll(reader)
+				if err != nil {
+					return fmt.Errorf("error reading response body: %v", err)
+				}
+
+				ibc := &relayer.IBCdata{}
+				if err = json.Unmarshal(b, &ibc); err != nil {
+					return fmt.Errorf("failed to unmarshal: %v ", err)
+				}
+
+				srcChainName := ibc.Chain1.ChainName
+				dstChainName := ibc.Chain2.ChainName
+
+				srcPathEnd := &relayer.PathEnd{
+					ChainID:      a.Config.Chains[srcChainName].ChainID(),
+					ClientID:     ibc.Chain1.ClientID,
+					ConnectionID: ibc.Chain1.ConnectionID,
 				}
+				dstPathEnd := &relayer.PathEnd{
+					ChainID:      a.Config.Chains[dstChainName].ChainID(),
+					ClientID:     ibc.Chain2.ClientID,
+					ConnectionID: ibc.Chain2.ConnectionID,
+				}
+				newPath := &relayer.Path{
+					Src: srcPathEnd,
+					Dst: dstPathEnd,
+				}
+				reader.Close()
+
+				if err = a.Config.AddPath(pthName, newPath); err != nil {
+					return fmt.Errorf("failed to add path %s: %w", pthName, err)
+				}
+				fmt.Fprintf(cmd.ErrOrStderr(), "added:  %s\n", pthName)
+
+			}
+
+			if err := a.OverwriteConfig(a.Config); err != nil {
+				return err
 			}
-			cleanupDir(localRepo)
 			return nil
+
 		},
 	}
-	return cmd
+	return OverwriteConfigFlag(a.Viper, cmd)
+}
+
+// used in pathsFetch command to find all possible combinations of paths for configured chains
+func permutations(arr []string) []string {
```

Comment:
Would you mind using https://pkg.go.dev/github.com/cespare/permute/v2 instead?

Reply:
This package seemed great, however I am unable to get the permutations outputed by Permute() into a slice of slices.

```go
	chain_list := []string{"Cat", "Dog", "Snake"}
	p := permute.Slice(chain_list)
	var permutations [][]string
	for p.Permute() {
		x := chain_list
		fmt.Println(x)
		permutations = append(permutations, x)
	}
	fmt.Println("PERMUTATIONS HERE: ", permutations)
```
```
[Cat Dog Snake]
[Dog Cat Snake]
[Snake Cat Dog]
[Cat Snake Dog]
[Dog Snake Cat]
[Snake Dog Cat]
PERMUTATIONS HERE:  [[Snake Dog Cat] [Snake Dog Cat] [Snake Dog Cat] [Snake Dog Cat] [Snake Dog Cat] [Snake Dog Cat]]
```
The printed slice is different than what gets appended to the array. 
I feel like I'm missing something simple here.

Reply:
Go slices are tricky because they reference to the same backing array even when you assign a new variable like `x`. You can think of slices as a "view" of an array. 

https://go.dev/blog/slices-intro
> re-slicing a slice doesn’t make a copy of the underlying array

In this case `chain_list` and `x` are sharing the same backing array which is why your `PERMUTATIONS HERE:` prints only the last combination of `Snake Dog Cat`. 

(Side: Note, I believe these are actually combinations, not permutations because order doesn't matter.)

A couple options that work:

```go
	chain_list := []string{"Cat", "Dog", "Snake"}
	p := permute.Slice(chain_list)
	var permutations [][]string
	for p.Permute() {
		fmt.Println(chain_list)
		x := make([]string, len(chain_list))
		copy(x, chain_list)
		permutations = append(permutations, x)
	}
	fmt.Println("PERMUTATIONS HERE: ", permutations)
```

Or less lines of codes:

```go
	chain_list := []string{"Cat", "Dog", "Snake"}
	p := permute.Slice(chain_list)
	var permutations [][]string
	for p.Permute() {
		fmt.Println(chain_list)
		permutations = append(permutations, append([]string{}, chain_list...))
	}
	fmt.Println("PERMUTATIONS HERE: ", permutations)
```

The solution is you must explicitly allocate a new array. 

Go's reason for this, AFAIK, is to conserve memory. Go won't allocate a new array unless it has to.

Fyi, I use this a lot: https://ueokande.github.io/go-slice-tricks/

Reply:
Thanks for the explanation here @DavidNix!
Got this working.